httpseverywhere.go

```diff
@@ -0,0 +1,262 @@
+	dec := gob.NewDecoder(buf)
+	targets := make(map[string]*Rules)
+	err := dec.Decode(&targets)
+	if err != nil {
+		log.Errorf("Could not decode: %v", err)
+		return nil, err
+	}
+
+	// The compiled regular expressions aren't serialized, so we have to manually
+	// compile them.
+	for _, v := range targets {
+		for _, r := range v.Rules {
+			r.from, _ = regexp.Compile(r.From)
+		}
+
+		for _, e := range v.Exclusions {
+			e.pattern, _ = regexp.Compile(e.Pattern)
+		}
+	}
+	h := &https{log: log, targets: targets}
+	return h.ToHTTPS, nil
+}
+
+// NewHTTPS creates a new ToHTTPS instance from a single rule set string.
+func NewHTTPS(rules string) ToHTTPS {
+	targets := make(map[string]*Rules)
+	AddRuleSet([]byte(rules), targets)
+	return &https{log: log, targets: targets}
+}
+
+// AddRuleSet adds the specified rule set to the map of targets.
+func AddRuleSet(rules []byte, targets map[string]*Rules) bool {
+	var r Ruleset
+	xml.Unmarshal(rules, &r)
+
+	// If the rule is turned off, ignore it.
+	if len(r.Off) > 0 {
+		return false
+	}
+
+	// We don't run on any platforms (aka Tor) that support mixed content, so
+	// ignore any rule that is mixedcontent-only.
+	if r.Platform == "mixedcontent" {
+		return false
+	}
+
+	rs, err := ruleSetToRules(r)
+	if err != nil {
+		return false
+	}
+
+	for _, target := range r.Target {
+		targets[target.Host] = rs
+	}
+	return true
+}
+
+func ruleSetToRules(set Ruleset) (*Rules, error) {
+	mod := make([]*rule, 0)
+	for _, r := range set.Rule {
+		// We ignore any rules that attempt to redirect to HTTP, as they would
+		// trigger mixed content in most cases (all cases in browsers that don't
+		// allow mixed content)?
+		if r.To == "http:" {
+			continue
+		}
+		f, err := regexp.Compile(r.From)
+		if err != nil {
+			log.Debugf("Could not compile regex: %v", err)
+			return &Rules{}, err
+		}
+		mod = append(mod, &rule{From: r.From, from: f, To: r.To})
+
+	}
+	exclude := make([]*exclusion, 0)
+	for _, e := range set.Exclusion {
+		p, err := regexp.Compile(e.Pattern)
+		if err != nil {
+			log.Debugf("Could not compile regex for exclusion: %v", err)
+			return &Rules{}, err
+		}
+		exclude = append(exclude, &exclusion{Pattern: e.Pattern, pattern: p})
+	}
+	return &Rules{Rules: mod, Exclusions: exclude}, nil
+}
+
+func (h *https) ToHTTPS(urlStr string) (string, bool) {
+	u, err := url.Parse(urlStr)
+	if err != nil {
+		log.Errorf("Could not parse domain %v", err)
+		return urlStr, false
+	}
+	domain := stripPort(u.Host)
+	if err != nil {
+		return urlStr, false
+	}
+
+	// We need to check for both the domain itself as well as the wildcard domain.
+
+	if rules, ok := h.targets[domain]; ok {
+		//h.log.Debugf("Got rules: %+v", rules)
+		return rules.ToHTTPS(urlStr)
+	}
+	if rules, ok := h.targets[stripTLD(domain)+"*"]; ok {
+		//h.log.Debugf("Got suffix rules: %+v", rules)
+		return rules.ToHTTPS(urlStr)
+	}
+	if rules, ok := h.targets["*."+stripSubdomains(domain)]; ok {
```

Comment:
OK, looks like it's 181 rules that truly have nested subdomains. It's super tempting to skip 'em to me in favor of linear time lookups, which would be similar to skipping the ~2000 rules we now skip that include backtracking in their regex (not supported in Go because it's inefficient and in pathological cases a security risk). Most of the ones that look like nested subdomains are really TLDs -- see https://www.publicsuffix.org/list/public_suffix_list.dat